server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -413,10 +413,9 @@ public void updateShardState(final ShardRouting newRouting,
 
             if (state == IndexShardState.POST_RECOVERY && newRouting.active()) {
                 assert currentRouting.active() == false : "we are in POST_RECOVERY, but our shard routing is active " + currentRouting;
-
-                if (newRouting.primary() && currentRouting.isRelocationTarget() == false) {
-                    replicationTracker.activatePrimaryMode(getLocalCheckpoint());
-                }
+                assert currentRouting.isRelocationTarget()  == false || currentRouting.primary() == false ||
```

Comment:
this assertion confuses me: What should it express?

Reply:
Agreed it's confusing. I tried to clarify by adding the assertion message. If it helps I can flip the boolean around:

```
(currentRouting.isRelocationTarget() && currentRouting.primary() && replicationTracker.isPrimaryMode() == false) == false 
```

but I think that might be just as confusing if not more. Let me know.

Reply:
let's keep as is, with the assertion message I think it's ok. I wonder if we should have an assertion at the end of this method to say something like "if we have an active primary shard that's not relocating, then the replication tracker is in primary mode".

Reply:
sounds good. will add.